src/hash.cpp

```diff
@@ -0,0 +1,156 @@
+/*
+ * #%L
+ * %%
+ * Copyright (C) 2018 BMW Car IT GmbH
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+#include "mococrw/hash.h"
+#include "mococrw/util.h"
+#include "mococrw/openssl_wrap.h"
+
+namespace mococrw
+{
+using namespace openssl;
+
+static const EVP_MD* _getDigestFn (DigestTypes digestType) {
+    switch (digestType) {
+        case DigestTypes::SHA256:
+            return lib::OpenSSLLib::SSL_EVP_sha256();
+        case DigestTypes::SHA512:
+            return lib::OpenSSLLib::SSL_EVP_sha512();
+        default:
+            throw std::runtime_error("Unknown digest type");
+    }
+}
+
+static EVP_MD_CTX _initCtx (DigestTypes digestType) {
+    const EVP_MD* digestFn = _getDigestFn(digestType);
+    EVP_MD_CTX digestCtx;
+    EVP_MD_CTX_init(&digestCtx);
+    EVP_DigestInit_ex(&digestCtx, digestFn, NULL);
+    return digestCtx;
+}
+
+static void _finalizeDigest (EVP_MD_CTX& digestCtx, uint8_t* outDigest, unsigned int &outLength) {
+    EVP_DigestFinal_ex(&digestCtx, outDigest, &outLength);
+    EVP_MD_CTX_cleanup(&digestCtx);
+}
+
+static void _hash(DigestTypes digestType, const uint8_t* message, const int &length, uint8_t* outDigest, unsigned int &outLength) {
```

Comment:
This function does no memory management at all and kind of outsources all the work to the caller. This is generally a rather C-style design and should be avoided in C++ as your callers have a tendency of doing things way too often. I would prefer a design that implements all functionality in a clean way in the class. The free functions should just use the class interface to implement their behavior. They are just convenience wrappers, not a high speed shortcut.

See also my other comments on output parameters.

Reply:
done